### PR TITLE
8297936: Use reachabilityFence to manage liveness in ClassUnload tests

### PR DESCRIPTION
**Pull request by afshin-zafari**
### Description
Instead of using static instances or using an object in print to keep it alive, `java.lang.ref.Reference.reachabilityFence(Object)` is used to keep an object alive and avoid class unloading.

### Tests
linux-x64-debug, macosx-aarch64-debug, mach5 tier1-5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297936](https://bugs.openjdk.org/browse/JDK-8297936): Use reachabilityFence to manage liveness in ClassUnload tests

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12552/head:pull/12552` \
`$ git checkout pull/12552`

Update a local copy of the PR: \
`$ git checkout pull/12552` \
`$ git pull https://git.openjdk.org/jdk pull/12552/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12552`

View PR using the GUI difftool: \
`$ git pr show -t 12552`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12552.diff">https://git.openjdk.org/jdk/pull/12552.diff</a>

</details>

### CODE REVIEW

**Review by coleenp — CHANGES_REQUESTED** [review]
I know -Xcomp shows the problem but I don't really want these tests to be run by default with -Xcomp.  Can you remove Xcomp?

File: test/hotspot/jtreg/runtime/ClassUnload/KeepAliveObject.java [hunk 1]
```diff
@@ -66,8 +65,8 @@ public static void main(String... args) throws Exception {
 
         ClassUnloadCommon.failIf(!isAlive, "should be alive");
     }
+    Reference.reachabilityFence(o);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ClassUnload/KeepAliveObject.java:68]: Please add comment:
```
// Don't let `o` get prematurely reclaimed by the GC.
```

File: test/hotspot/jtreg/runtime/ClassUnload/UnloadTest.java [hunk 2]
```diff
@@ -49,11 +50,10 @@
  */
 public class UnloadTest {
     // Using a global static field to keep the object live in -Xcomp mode.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ClassUnload/UnloadTest.java:52]: This comment needs deleting now.

File: test/hotspot/jtreg/runtime/ClassUnload/UnloadTest.java [hunk 3]
```diff
@@ -76,6 +76,7 @@ private static void test_unload_instance_klass() throws Exception {
 
         ClassUnloadCommon.failIf(!wb.isClassAlive(className), "should still be live");
 
+        Reference.reachabilityFence(o);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ClassUnload/UnloadTest.java:79]: Again add a comment please.

File: test/hotspot/jtreg/runtime/ClassUnload/UnloadTest.java [hunk 4]
```diff
@@ -104,6 +105,7 @@ private static void test_unload_obj_array_klass() throws Exception {
         ClassUnloadCommon.triggerUnloading();
         ClassUnloadCommon.failIf(!wb.isClassAlive(className), "should still be live");
 
+        Reference.reachabilityFence(o);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ClassUnload/UnloadTest.java:108]: Again please add comment.

File: test/hotspot/jtreg/runtime/ClassUnload/ConstantPoolDependsTest.java [hunk 5]
```diff
@@ -32,12 +32,13 @@
  * @build jdk.test.whitebox.WhiteBox
  * @compile p2/c2.java MyDiffClassLoader.java
  * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
- * @run main/othervm -Xbootclasspath/a:. -Xmn8m -XX:+UnlockDiagnosticVMOptions -Xlog:class+unload -XX:+WhiteBoxAPI ConstantPoolDependsTest
+ * @run main/othervm -Xbootclasspath/a:. -Xmn8m -XX:+UnlockDiagnosticVMOptions -Xlog:class+unload -Xcomp -XX:+WhiteBoxAPI ConstantPoolDependsTest
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ClassUnload/ConstantPoolDependsTest.java:35]: As Coleen states, do not force the tests to run in Xcomp mode.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Comment by afshin-zafari** [comment]
> I know -Xcomp shows the problem but I don't really want these tests to be run by default with -Xcomp. Can you remove Xcomp?

done.

**Review by dholmes-ora — APPROVED** [review]
Updates look good. Thanks.

**Pull request closed without merging**